File: src/WorkBC.Web/Services/RecommendedJobsService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;
using WorkBC.Data;
using WorkBC.Data.Model.JobBoard;
using WorkBC.ElasticSearch.Models.Filters;
using WorkBC.ElasticSearch.Models.JobAttributes;
using WorkBC.ElasticSearch.Models.Results;
using WorkBC.ElasticSearch.Search.Queries;
using WorkBC.Shared.Services;
using WorkBC.Web.Helpers;
using WorkBC.Web.Models;
using Constants = WorkBC.ElasticSearch.Search.Boosts.RecommendedJobsBoost;

namespace WorkBC.Web.Services
{
    public interface IRecommendedJobsService
    {
        Task<RecommendedJobsResultModel> GetRecommendedJobs(RecommendedJobsFilters filters, AccountCriteria accountCriteria);

        Task<long> GetRecommendedJobCount(AccountCriteria accountCriteria);

        Task<AccountCriteria> GetUserAccountCriteria(JobSeeker jobSeeker);
    }

    public class RecommendedJobsService : IRecommendedJobsService
    {
        private const int BritishColumbia = 2;
        private readonly IConfiguration _configuration;
        private readonly JobBoardContext _context;
        private readonly IGeocodingService _geocodingService;
        private readonly ILogger<RecommendedJobsQuery> _loggerElasticSearch;
        private readonly ViewCountService _viewCountService;

        public RecommendedJobsService(JobBoardContext context, IGeocodingService geocodingService,
            IConfiguration configuration,
            ILogger<RecommendedJobsQuery> loggerElasticSearch, CacheService cacheService)
        {
            _context = context;
            _geocodingService = geocodingService;
            _configuration = configuration;
            _loggerElasticSearch = loggerElasticSearch;
            _viewCountService = new ViewCountService(cacheService, context);
        }

        private RecommendedJobsFilters ElasticFilters { get; set; }
        private AccountCriteria AccountCriteria { get; set; }

        /// <summary>
        ///     Gets recommended jobs for a user
        /// </summary>
        public async Task<RecommendedJobsResultModel> GetRecommendedJobs(RecommendedJobsFilters filters, AccountCriteria accountCriteria)
        {
            // add the job seeker data to the filter that was posted from the ajax call
            BuildFilter(filters, accountCriteria);

            //Get search results from Elastic search
            var esq = new RecommendedJobsQuery(_configuration, ElasticFilters);
            ElasticSearchResponse results = await esq.GetSearchResults();

            //Build the object that we will return to the client
            var model = new RecommendedJobsResultModel
            {
                PageNumber = ElasticFilters.Page,
                PageSize = ElasticFilters.PageSize
            };

            if (results != null)
            {
                if (results.Hits?.HitsHits != null)
                {
                    model.Count = results.Hits.Total.Value ?? 0;

                    // add reasons why the job was recommended
                    IEnumerable<Source> sources = results.Hits.HitsHits.Select(hit => hit.Source);
                    decimal?[] scores = results.Hits.HitsHits.Select(hit => hit.Score).ToArray();

                    model.Result = AddReasons(sources, scores);

                    //Get the number of views for each job and get the data back
                    model.Result = await _viewCountService.GetJobViews(model.Result);
                }
                else
                {
                    model.Result = new Source[0];
                }
            }

            return model;
        }

        /// <summary>
        ///     Gets recommended jobs for a user
        /// </summary>
        public async Task<long> GetRecommendedJobCount(AccountCriteria accountCriteria)
        {
            var filters = new RecommendedJobsFilters
            {
                FilterSavedJobNocs = false,
                FilterSavedJobTitles = false,
                FilterSavedJobEmployers = false,
                FilterIsApprentice = false,
                FilterJobSeekerCity = false,
                FilterIsIndigenous = false,
                FilterIsMatureWorkers = false,
                FilterIsNewcomers = false,
                FilterIsPeopleWithDisabilities = false,
                FilterIsStudents = false,
                FilterIsVeterans = false,
                FilterIsVisibleMinority = false,
                FilterIsYouth = false,
                PageSize = 0,
                Page = 1
            };

            // add the job seeker data to the filter that was posted from the ajax call
            BuildFilter(filters, accountCriteria);

            //Get search results from Elastic search
            var esq = new RecommendedJobsQuery(_configuration, ElasticFilters);
            ElasticSearchResponse results = await esq.GetSearchResults();

            if (results?.Hits?.HitsHits != null)
            {
                return results.Hits.Total.Value ?? 0;
            }

            return 0;
        }

        /// <summary>
        ///     Gets all the raw data needed to do a recommended jobs query in ElasticSearch for a user
        /// </summary>
        public async Task<AccountCriteria> GetUserAccountCriteria(JobSeeker jobSeeker)
        {
            // get the 200 most recent saved jobs for the user
            List<Job> savedJobs = await (from sj in _context.SavedJobs
                                         join j in _context.Jobs on sj.JobId equals j.JobId
                                         where sj.AspNetUserId == jobSeeker.Id && !sj.IsDeleted
                                         orderby sj.Id descending
                                         select new Job
                                         {
                                             JobId = j.JobId,
                                             City = j.City,
                                             DatePosted = j.DatePosted,
                                             //NocCodeId = j.NocCodeId,
                                             NocCodeId2021 = j.NocCodeId2021,
                                             EmployerName = j.EmployerName.ToLower(), // lowercase names for grouping
                                             Title = j.Title.ToLower(), // lowercase names for grouping
                                             IsActive = j.IsActive
                                         }).Take(Constants.MaxSavedJobs).Distinct().ToListAsync();

            // get cityLabel, latitude, longitude
            string cityLabel = await GetUserLocationInfo(jobSeeker);

            // get the user's groups (e.g. Indigenous)
            JobSeekerFlags userFlags =
                await _context.JobSeekerFlags.FirstOrDefaultAsync(f => f.JobSeeker.Id == jobSeeker.Id);

            //// group the saved jobs by noc code
            //Dictionary<short, int> nocCodes = savedJobs
            //    .GroupBy(j => j.NocCodeId ?? 0)
            //    .Select(g => new {Term = g.Key, Count = g.Count()})
            //    .OrderByDescending(a => a.Count)
            //    .ToDictionary(k => k.Term, v => v.Count);

            // group the saved jobs by noc code 2021
            Dictionary<int, int> nocCodes2021 = savedJobs
                .GroupBy(j => j.NocCodeId2021 ?? 0)
                .Select(g => new { Term = g.Key, Count = g.Count() })
                .OrderByDescending(a => a.Count)
                .ToDictionary(k => k.Term, v => v.Count);

            // group the saved jobs by Employer
            Dictionary<string, int> employers = savedJobs
                .GroupBy(j => j.EmployerName)
                .Select(g => new {Term = g.Key, Count = g.Count()})
                .OrderByDescending(a => a.Count)
                .ToDictionary(k => k.Term, v => v.Count);

            // group the saved jobs by title
            Dictionary<string, int> titles = savedJobs
                .GroupBy(j => j.Title)
                .Select(g => new {Term = g.Key, Count = g.Count()})
                .OrderByDescending(a => a.Count)
                .ToDictionary(k => k.Term, v => v.Count);

            // saved job id's should be ignored
            string[] savedJobIds = savedJobs.Where(j => j.IsActive).Select(j => j.JobId.ToString()).ToArray();

            // return all the parameters
            return new AccountCriteria
            {
                InBritishColumbia = jobSeeker.ProvinceId == BritishColumbia,
                City = cityLabel,
                SavedJobIds = savedJobIds,
                UserFlags = userFlags,
                //NocCodes = nocCodes,
                NocCodes2021 = nocCodes2021,
                Employers = employers,
                Titles = titles,
            };
        }

        /// <summary>
        ///     Adds data from the user's account to the filter that was submitted from the
        ///     Angular app.
        /// </summary>
        private void BuildFilter(RecommendedJobsFilters filter, AccountCriteria accountCriteria)
        {
            AccountCriteria = accountCriteria;

            JobSeekerFlags flags = AccountCriteria.UserFlags;

            filter.IsApprentice = flags.IsApprentice;
            filter.IsIndigenous = flags.IsIndigenousPerson;
            filter.IsMatureWorkers = flags.IsMatureWorker;
            filter.IsNewcomers = flags.IsNewImmigrant;
            filter.IsPeopleWithDisabilities = flags.IsPersonWithDisability;
            filter.IsStudents = flags.IsStudent;
            filter.IsVeterans = flags.IsVeteran;
            filter.IsVisibleMinority = flags.IsVisibleMinority;
            filter.IsYouth = flags.IsYouth;

            filter.FilterIsApprentice = filter.FilterIsApprentice && flags.IsApprentice;
            filter.FilterIsIndigenous = filter.FilterIsIndigenous && flags.IsIndigenousPerson;
            filter.FilterIsMatureWorkers = filter.FilterIsMatureWorkers && flags.IsMatureWorker;
            filter.FilterIsNewcomers = filter.FilterIsNewcomers && flags.IsNewImmigrant;
            filter.FilterIsPeopleWithDisabilities = filter.FilterIsPeopleWithDisabilities && flags.IsPersonWithDisability;
            filter.FilterIsStudents = filter.FilterIsStudents && flags.IsStudent;
            filter.FilterIsVeterans = filter.FilterIsVeterans && flags.IsVeteran;
            filter.FilterIsVisibleMinority = filter.FilterIsVisibleMinority && flags.IsVisibleMinority;
            filter.FilterIsYouth = filter.FilterIsYouth && flags.IsYouth;

            if (AccountCriteria.InBritishColumbia)
            {
                filter.City = AccountCriteria.City;
            }

            //filter.NocCodes = AccountCriteria.NocCodes;
            filter.NocCodes2021 = AccountCriteria.NocCodes2021;
            filter.Titles = AccountCriteria.Titles;
            filter.Employers = AccountCriteria.Employers;

            filter.IgnoreJobIdList = AccountCriteria.SavedJobIds;

            ElasticFilters = filter;
        }

        /// <summary>
        ///     Gets the city label for users from BC.  City label is usually just the city name,
        ///     but in cases where there are two places in BC with the same name it also contains
        ///     the region name.
        /// </summary>
        private async Task<string> GetUserLocationInfo(JobSeeker jobSeeker)
        {
            string cityLabel;
            if (jobSeeker.ProvinceId == BritishColumbia)
            {
                cityLabel = (await _context.Locations
                        .FirstOrDefaultAsync(l => l.LocationId == jobSeeker.LocationId))?.Label ?? "_NO_CITY_";
            }
            else
            {
                cityLabel = null;

            }

            return cityLabel;
        }

        /// <summary>
        ///     Adds the Reason to each job (a sentence describing why the job was recommended)
        ///     Also strips out fields that are only needed to construct the reason sentence from
        ///     the final json output.
        /// </summary>
        private Source[] AddReasons(IEnumerable<Source> results, decimal?[] scores)
        {
            var newSource = new List<Source>();

            int i = 0;

            foreach (Source result in results)
            {
                var s = new Source
                {
                    JobId = result.JobId,
                    City = result.City,
                    DatePosted = result.DatePosted,
                    EmployerName = result.EmployerName,
                    ExpireDate = result.ExpireDate,
                    HoursOfWork = result.HoursOfWork,
                    IsFederalJob = result.IsFederalJob,
                    LastUpdated = result.LastUpdated,
                    //Noc = result.Noc,
                    Noc2021 = result.Noc2021,
                    PeriodOfEmployment = result.PeriodOfEmployment,
                    SalarySummary = result.SalarySummary,
                    Title = result.Title,
                    Reason = GetReason(result),
                    ExternalSource = result.ExternalSource,
                    Score = scores[i]
                };

                newSource.Add(s);
                i++;
            }

            return newSource.ToArray();
        }

        /// <summary>
        ///     Gets a sentence describing why a job was recommended
        /// </summary>
        public string GetReason(Source result)
        {
            //return "Recommended ";

            var reasons = new List<string>();

            if (result.Noc2021 != null)
            {                
                int noc= Int32.TryParse(result.Noc2021, out noc) ? noc : 0;
                if (AccountCriteria.NocCodes2021.ContainsKey(noc))
                {
                    int nocCount = AccountCriteria.NocCodes2021[noc];
                    reasons.Add($"based on having the same NOC code as {nocCount.GetWord()} of your saved jobs");
                }
            }

            string title = result.Title.ToLower();
            if (AccountCriteria.Titles.ContainsKey(title))
            {
                int titleCount = AccountCriteria.Titles[title];
                reasons.Add($"based on having the same job title as {titleCount.GetWord()} of your saved jobs");
            }

            string employer = result.EmployerName.ToLower();
            if (AccountCriteria.Employers.ContainsKey(employer))
            {
                int employerCount = AccountCriteria.Employers[employer];
                reasons.Add($"based on having the same employer as {employerCount.GetWord()} of your saved jobs");
            }

            if (!string.IsNullOrWhiteSpace(AccountCriteria?.City) && result.City.Contains(AccountCriteria.City))
            {
                reasons.Add($"based on your location of {AccountCriteria.City} in your account settings");
            }

            var groups = new List<string>();

            if (ElasticFilters.IsNewcomers && (result.IsNewcomer ?? false))
            {
                groups.Add("a newcomer to B.C.");
            }

            if (ElasticFilters.IsStudents && (result.IsStudent ?? false))
            {
                groups.Add("a student");
            }

            if (ElasticFilters.IsVeterans && (result.IsVeteran ?? false))
            {
                groups.Add("a veteran of the Canadian Armed Forces");
            }

            if (ElasticFilters.IsVisibleMinority && (result.IsVismin ?? false))
            {
                groups.Add("a visible minority");
            }

            if (ElasticFilters.IsYouth && (result.IsYouth ?? false))
            {
                groups.Add("a youth");
            }

            if (ElasticFilters.IsPeopleWithDisabilities && (result.IsDisability ?? false))
            {
                groups.Add("a person with a disability");
            }

            if (ElasticFilters.IsIndigenous && (result.IsAboriginal ?? false))
            {
                groups.Add("an Indigenous person");
            }

            if (ElasticFilters.IsApprentice && (result.IsApprentice ?? false))
            {
                groups.Add("an apprentice");
            }

            if (ElasticFilters.IsMatureWorkers && (result.IsMatureWorker ?? false))
            {
                groups.Add("a mature worker");
            }

            var groupsText = "";
            if (groups.Count == 1)
            {
                groupsText = groups[0];
            }
            else if (groups.Count > 1)
            {
                groupsText = $"{string.Join(", ", groups.Take(groups.Count - 1))} or {groups[^1]}";
            }

            if (groups.Any())
            {
                reasons.Add(
                    $"because this job posting encourages applications from those who identify as {groupsText}");
            }

            if (reasons.Count <= 0)
            {
                return "";
            }

            string reason = reasons.Count == 1
                ? $"Recommended {reasons[0]}"
                : $"Recommended {string.Join(", ", reasons.Take(reasons.Count - 1))} and {reasons[^1]}";

            return reason.EndsWith(".") ? reason : $"{reason}.";

        }
    }

    public class AccountCriteria
    {
        public Dictionary<string, int> Titles { get; set; }
        //public Dictionary<short, int> NocCodes { get; set; }
        public Dictionary<int, int> NocCodes2021 { get; set; }
        public Dictionary<string, int> Employers { get; set; }
        public JobSeekerFlags UserFlags { get; set; }
        public string[] SavedJobIds { get; set; }
        public bool InBritishColumbia { get; set; }
        public string City { get; set; }
    }
}

File: jobboard-laravel/app/Services/JobSeeker/RecommendedJobsService.php
<?php

namespace App\Services\JobSeeker;

use App\Models\JobSeeker;
use App\Models\SavedJob;
use App\Search\Queries\RecommendedJobsQuery;
use App\Search\Results\JobResult;
use OpenSearch\Client;

final class RecommendedJobsService
{
    /** @var array<string, string> */
    private const SEEKER_FLAG_TO_INDEX_FIELD = [
        'IsApprentice' => 'IsApprentice',
        'IsVeteran' => 'IsVeteran',
        'IsIndigenousPerson' => 'IsAboriginal',
        'IsMatureWorker' => 'IsMatureWorker',
        'IsNewImmigrant' => 'IsNewcomer',
        'IsPersonWithDisability' => 'IsDisability',
        'IsStudent' => 'IsStudent',
        'IsVisibleMinority' => 'IsVismin',
        'IsYouth' => 'IsYouth',
    ];

    /** @var array<string, string> */
    private const EQUITY_REASON_LABELS = [
        'IsApprentice' => 'an apprentice',
        'IsVeteran' => 'a veteran',
        'IsAboriginal' => 'an Indigenous person',
        'IsMatureWorker' => 'a mature worker',
        'IsNewcomer' => 'a newcomer',
        'IsDisability' => 'a person with a disability',
        'IsStudent' => 'a student',
        'IsVismin' => 'a visible minority',
        'IsYouth' => 'youth',
    ];

    public function __construct(private Client $client) {}

    public function aggregateSignals(JobSeeker $jobSeeker): RecommendedJobSignals
    {
        $rows = SavedJob::query()
            ->leftJoin('Jobs', 'Jobs.JobId', '=', 'SavedJobs.JobId')
            ->where('SavedJobs.AspNetUserId', (string) $jobSeeker->Id)
            ->orderByDesc('SavedJobs.DateSaved')
            ->limit(RecommendedJobsQuery::MAX_SAVED_JOBS)
            ->get([
                'SavedJobs.JobId',
                'Jobs.NocCodeId2021',
                'Jobs.EmployerName',
                'Jobs.Title',
            ]);

        $nocCounts = [];
        $employerCounts = [];
        $titleCounts = [];
        $savedJobIds = [];

        foreach ($rows as $row) {
            $jobId = trim((string) ($row->JobId ?? ''));
            if ($jobId !== '') {
                $savedJobIds[$jobId] = true;
            }

            $noc = (int) ($row->NocCodeId2021 ?? 0);
            $nocCounts[$noc] = ($nocCounts[$noc] ?? 0) + 1;

            $employer = $this->normalizeText($row->EmployerName ?? null);
            if ($employer !== '') {
                $employerCounts[$employer] = ($employerCounts[$employer] ?? 0) + 1;
            }

            $title = $this->normalizeText($row->Title ?? null);
            if ($title !== '') {
                $titleCounts[$title] = ($titleCounts[$title] ?? 0) + 1;
            }
        }

        $equityFields = [];
        $flags = $jobSeeker->flags()->first();

        if ($flags !== null) {
            foreach (self::SEEKER_FLAG_TO_INDEX_FIELD as $seekerField => $indexField) {
                if ((bool) ($flags->{$seekerField} ?? false)) {
                    $equityFields[] = $indexField;
                }
            }
        }

        return new RecommendedJobSignals(
            savedJobIds: array_values(array_keys($savedJobIds)),
            nocCounts: $nocCounts,
            employerCounts: $employerCounts,
            titleCounts: $titleCounts,
            city: $this->normalizeText($jobSeeker->City ?? null),
            equityFields: $equityFields,
        );
    }

    /**
     * @return array{signals: RecommendedJobSignals, total: int, jobs: list<array<string, mixed>>}
     */
    public function recommendationsFor(JobSeeker $jobSeeker, int $page = 1, int $pageSize = 20): array
    {
        $signals = $this->aggregateSignals($jobSeeker);

        if (! $signals->hasSavedJobs()) {
            return ['signals' => $signals, 'total' => 0, 'jobs' => []];
        }

        $response = $this->client->search([
            'index' => $this->index(),
            'body' => (new RecommendedJobsQuery(
                nocCounts: $signals->nocCounts,
                employerCounts: $signals->employerCounts,
                titleCounts: $signals->titleCounts,
                city: $signals->city,
                equityFields: $signals->equityFields,
                ignoreJobIds: $signals->savedJobIds,
                page: $page,
                pageSize: $pageSize,
            ))->build(),
        ]);

        $hits = $response['hits']['hits'] ?? [];
        $total = (int) ($response['hits']['total']['value'] ?? 0);

        $jobs = [];
        foreach ($hits as $hit) {
            $source = is_array($hit['_source'] ?? null) ? $hit['_source'] : [];
            $job = JobResult::fromSource($source)->toArray();
            $job['Score'] = (float) ($hit['_score'] ?? 0.0);
            $job['Reason'] = $this->reasonFor($source, $signals);
            $jobs[] = $job;
        }

        return [
            'signals' => $signals,
            'total' => $total,
            'jobs' => $jobs,
        ];
    }

    public function recommendedCountFor(JobSeeker $jobSeeker): int
    {
        $signals = $this->aggregateSignals($jobSeeker);

        if (! $signals->hasSavedJobs()) {
            return 0;
        }

        $response = $this->client->search([
            'index' => $this->index(),
            'body' => (new RecommendedJobsQuery(
                nocCounts: $signals->nocCounts,
                employerCounts: $signals->employerCounts,
                titleCounts: $signals->titleCounts,
                city: $signals->city,
                equityFields: $signals->equityFields,
                ignoreJobIds: $signals->savedJobIds,
                page: 1,
                pageSize: 0,
            ))->build(),
        ]);

        return (int) ($response['hits']['total']['value'] ?? 0);
    }

    /**
     * @param  array<string, mixed>  $source
     */
    private function reasonFor(array $source, RecommendedJobSignals $signals): string
    {
        $clauses = [];

        $noc = (int) round((float) ($source['Noc2021'] ?? 0));
        $nocCount = $signals->nocCounts[$noc] ?? 0;
        if ($nocCount > 0) {
            $clauses[] = sprintf(
                'having the same NOC code as %s of your saved jobs',
                $this->countWord($nocCount),
            );
        }

        $employer = $this->normalizeText($source['EmployerName'] ?? null);
        $employerCount = $employer !== '' ? ($signals->employerCounts[$employer] ?? 0) : 0;
        if ($employerCount > 0) {
            $clauses[] = sprintf(
                'having the same employer as %s of your saved jobs',
                $this->countWord($employerCount),
            );
        }

        $title = $this->normalizeText($source['Title'] ?? null);
        $titleCount = $title !== '' ? ($signals->titleCounts[$title] ?? 0) : 0;
        if ($titleCount > 0) {
            $clauses[] = sprintf(
                'having the same job title as %s of your saved jobs',
                $this->countWord($titleCount),
            );
        }

        if ($signals->city !== null && $signals->city !== '') {
            $cities = $this->normalizeCities($source['City'] ?? null);
            if (in_array($signals->city, $cities, true)) {
                $clauses[] = 'being in your city';
            }
        }

        $equityMatches = [];
        foreach ($signals->equityFields as $field) {
            if (! array_key_exists($field, self::EQUITY_REASON_LABELS)) {
                continue;
            }

            if ((bool) ($source[$field] ?? false)) {
                $equityMatches[] = self::EQUITY_REASON_LABELS[$field];
            }
        }

        if ($clauses === [] && $equityMatches === []) {
            return 'Recommended based on your saved jobs and profile.';
        }

        $reason = 'Recommended based on '.implode(', ', $clauses).'.';

        if ($equityMatches !== []) {
            $reason .= ' This job also matches your profile as '.$this->joinWithAnd($equityMatches).'.';
        }

        return $reason;
    }

    /**
     * @return list<string>
     */
    private function normalizeCities(mixed $city): array
    {
        if (is_array($city)) {
            $cities = array_map(fn (mixed $value): string => $this->normalizeText($value), $city);

            return array_values(array_filter($cities, static fn (string $value): bool => $value !== ''));
        }

        $normalized = $this->normalizeText($city);

        return $normalized === '' ? [] : [$normalized];
    }

    private function normalizeText(mixed $value): string
    {
        $text = trim((string) ($value ?? ''));

        return $text === '' ? '' : mb_strtolower($text, 'UTF-8');
    }

    private function countWord(int $count): string
    {
        return match ($count) {
            1 => 'one',
            2 => 'two',
            3 => 'three',
            4 => 'four',
            5 => 'five',
            6 => 'six',
            7 => 'seven',
            8 => 'eight',
            9 => 'nine',
            10 => 'ten',
            default => (string) $count,
        };
    }

    /**
     * @param  list<string>  $parts
     */
    private function joinWithAnd(array $parts): string
    {
        $count = count($parts);

        if ($count <= 1) {
            return $parts[0] ?? '';
        }

        if ($count === 2) {
            return $parts[0].' and '.$parts[1];
        }

        $last = array_pop($parts);

        return implode(', ', $parts).', and '.$last;
    }

    private function index(): string
    {
        $key = app()->getLocale() === 'fr' ? 'fr' : 'en';

        return (string) config("opensearch.indexes.{$key}", config('opensearch.indexes.en'));
    }
}
